File: SRP/Ruby/não aplicado/pedido.rb
require 'ostruct'
require 'net/smtp'

class FormasPagamento
    CARTAOCREDITO = 1   
end

class Pedido
    def comprar(carrinho, detalhesDoPagamento, deveNotificaCliente)
        cobrarCartao(detalhesDoPagamento, carrinho) if detalhesDoPagamento.formaPagamento == FormasPagamento::CARTAOCREDITO
        
        reservaEstoque(carrinho)

        notificarCliente(carrinho) if deveNotificaCliente
    end 

    def cobrarCartao(detalhesDoPagamento, carrinho)
        gatewayPagamento = GatewayPagamento.new
        
        gatewayPagamento.credencial = "bla"
        gatewayPagamento.numeroCartao = detalhesDoPagamento.numeroCartao
        gatewayPagamento.mesExpiracao = detalhesDoPagamento.mesExpiracao
        gatewayPagamento.anoExpiracao = detalhesDoPagamento.anoExpiracao
        gatewayPagamento.nomeNoCartao = detalhesDoPagamento.nomeNoCartao
        gatewayPagamento.valor = carrinho.valorTotal

        gatewayPagamento.cobrar      
    end

    def reservaEstoque(carrinho)
        carrinho.itens.each do |item|
            sistemaEstoque = SistemaEstoque.new
            sistemaEstoque.reservar(item.Id, item.Quantidade)
        end     
    end

    def notificarCliente(carrinho)
        email = carrinho.emailCliente
        if !email.to_s.empty?
            message = <<MESSAGE_END
From: Loja <pedido@loja.com>
To: A Test User #{email}
Subject: SMTP e-mail test

This is a test e-mail message.
MESSAGE_END

            Net::SMTP.start('localhost') do |smtp|
                smtp.send_message message, 'pedido@loja.com', email
            end
        end
    end
end

File: SRP/C#/não aplicado/Pedido.cs
public enum FormasPagamento
{
    CARTAOCREDITO
}

public class Pedido
{
    public void Comprar(Carrinho carrinho, DetalhesPagamento detalhesPagamento, bool deveNotificarCliente)
    {
        if(detalhesPagamento.FormaPagamento == FormasPagamento.CARTAOCREDITO)
            CobrarCartao(detalhesDoPagamento, carrinho);
            
        ReservarEstoque(carrinho);
        
        if(deveNotificarCliente) 
            NotificarCliente(carrinho);
    }
    
    
    private void CobrarCartao(DetalhesPagamento detalhesDoPagamento, Carrinho carrinho)
    {
        var gatewayPagamento = new GatewayPagamento();
        
        gatewayPagamento.credencial   = "bla";
        gatewayPagamento.MumeroCartao = detalhesDoPagamento.MumeroCartao;
        gatewayPagamento.MesExpiracao = detalhesDoPagamento.MesExpiracao;
        gatewayPagamento.AnoExpiracao = detalhesDoPagamento.AnoExpiracao;
        gatewayPagamento.NomeNoCartao = detalhesDoPagamento.NomeNoCartao;
        gatewayPagamento.Valor        = carrinho.ValorTotal;
        gatewayPagamento.Cobrar();
    }
    
    private void ReservarEstoque(Carrinho carrinho)
    {
        foreach (var item in carrinho.Itens)
        {
            var sistemaEstoque = new SistemaEstoque();
            sistemaEstoque.Reservar(item.Id, item.Quantidade);
        }
    }
    
    private void NotificarCliente(Carrinho carrinho)
    {
        var emailCliente = carrinho.EmailCliente;
        
        if(string.IsNullOrWhiteSpace(emailCliente))
            return;
            
        using(var email = new MailMessage("pedido@loja.com", emailCliente))
        using(var smtp  = new SmtpClient("localhost"))
        {
            email.Body = "This is a test e-mail message.";
            email.Subject("SMTP e-mail test");
            smtp.Send(email);
        }
    }   
}